segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ReadOnlySegmentContainer.java

```diff
@@ -0,0 +1,192 @@
+    private final SegmentStateMapper segmentStateMapper;
+    private final ReadOnlyStorage storage;
+    private final ScheduledExecutorService executor;
+    private final AtomicBoolean closed;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the ReadOnlySegmentContainer class.
+     *
+     * @param storageFactory A StorageFactory used to create Storage adapters.
+     * @param executor       An Executor to use for async operations.
+     */
+    ReadOnlySegmentContainer(StorageFactory storageFactory, ScheduledExecutorService executor) {
+        Preconditions.checkNotNull(storageFactory, "storageFactory");
+        this.executor = Preconditions.checkNotNull(executor, "executor");
+        Storage writableStorage = storageFactory.createStorageAdapter();
+        this.storage = writableStorage;
+        this.stateStore = new SegmentStateStore(writableStorage, this.executor);
+        this.segmentStateMapper = new SegmentStateMapper(this.stateStore, writableStorage);
+        this.closed = new AtomicBoolean();
+    }
+
+    //endregion
+
+    //region AutoCloseable Implementation
+
+    @Override
+    public void close() {
+        if (this.closed.compareAndSet(false, true)) {
+            Futures.await(Services.stopAsync(this, this.executor));
+            this.storage.close();
+            log.info("Closed.");
+        }
+    }
+
+    //endregion
+
+    //region AbstractIdleService Implementation
+
+    @Override
+    protected Executor executor() {
+        return this.executor;
+    }
+
+    @Override
+    protected void startUp() {
+        this.storage.initialize(CONTAINER_EPOCH);
+        log.info("Started.");
+    }
+
+    @Override
+    protected void shutDown() {
+        log.info("Stopped.");
+    }
+
+    //endregion
+
+    // SegmentContainer Implementation
+
+    @Override
+    public int getId() {
+        return CONTAINER_ID;
+    }
+
+    @Override
+    public CompletableFuture<ReadResult> read(String streamSegmentName, long offset, int maxLength, Duration timeout) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        TimeoutTimer timer = new TimeoutTimer(timeout);
+        return getStreamSegmentInfo(streamSegmentName, false, timer.getRemaining())
+                .thenApply(si -> StreamSegmentStorageReader.read(si, offset, maxLength, MAX_READ_AT_ONCE_BYTES, this.storage));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> getStreamSegmentInfo(String streamSegmentName, boolean waitForPendingOps, Duration timeout) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        return this.segmentStateMapper.getSegmentInfoFromStorage(streamSegmentName, timeout);
+    }
+
+
+    //endregion
+
+    //region Unsupported Operations
+
+    @Override
+    public Collection<SegmentProperties> getActiveSegments() {
+        throw new UnsupportedOperationException("getActiveSegments is not supported on " + getClass().getSimpleName());
```

Comment:
Should be calling `unsupported("getActiveSegments");`

Reply:
This method does not return a `CompletableFuture`, so I can't use that.

Reply:
Ok, makes sense. I suppose the alternative would be to have a second method with different signature, which might not be worth doing for a single use, although might be slightly better for symmetry. I'm fine either way.